File: Python3/0395. Longest Substring with At Least K Repeating Characters.py
class Solution:
    def longestSubstring(self, s: str, k: int) -> int:
        res = 0
        for t in range(1,len(set(s))+1):
            l = 0
            r = 0
            letter = collections.defaultdict(int)
            total_type = 0

            while r<len(s):
                letter[s[r]] += 1
                if(letter[s[r]]==1):
                    total_type += 1
                while total_type > t:       
                    letter[s[l]] -= 1
                    if letter[s[l]] == 0:
                        total_type -= 1
                        del letter[s[l]]
                    l += 1
                else:
                    if(min(letter.values())>=k):        
                        res = max(res,r-l+1)
                r += 1

        return res

File: C++/0395. Longest Substring with At Least K Repeating Characters.cpp
class Solution {
public:
    int longestSubstring(string s, int k) {
        int res = 0;
        for(int t=1;t<=26;++t){
            int l = 0;
            int r = 0;
            vector<int> letter(26,(1<<10));
            int len = 0;
        
            while(r<s.size()){
                
                if((letter[s[r]-'a']) == (1<<10)){
                    len++;
                    letter[s[r]-'a'] = 1;
                }
                else{
                    letter[s[r]-'a']++;
                }

                while(len>t){
                    letter[s[l]-'a'] --;
                    if(letter[s[l]-'a']==0){
                        len--;
                        letter[s[l]-'a'] = (1<<10);
                    }
                    l++;
                }

                if(*min_element(letter.begin(),letter.end())>=k){
                    res = max(res,r-l+1);
                }
                r++;
            }
        }
        return res;
    }
};
